controlpanel/api/models/tool.py

```diff
@@ -0,0 +1,131 @@
+import logging
+import secrets
+
+from django.conf import settings
+from django.contrib.postgres.fields import JSONField
+import django.core.exceptions
+from django.db import models
+from django.db.models import Q
+from django_extensions.db.models import TimeStampedModel
+
+from controlpanel.api import cluster
+
+
+log = logging.getLogger(__name__)
+
+
+class Tool(TimeStampedModel):
+    """
+    Represents a software tool available to users of the platform. An instance
+    of Tool is an item in the Software Catalogue - not a user's deployed
+    instance of a tool.
+    """
+
+    description = models.TextField(blank=True)
+    chart_name = models.CharField(max_length=100, blank=False)
+    name = models.CharField(max_length=100, blank=False)
+    values = JSONField(default=dict)
+    version = models.CharField(max_length=100, blank=False)
+
+    class Meta(TimeStampedModel.Meta):
+        db_table = "control_panel_api_tool"
+        ordering = ('name',)
+
+    def __repr__(self):
+        return f'<Tool: {self.chart_name} {self.version}>'
+
+    def url(self, user):
+        return f'https://{user.slug}-{self.chart_name}.{settings.TOOLS_DOMAIN}/'
+
+
+class ToolDeploymentManager:
+    """
+    Emulates a Django model manager
+    """
+
+    def create(self, *args, **kwargs):
+        tool_deployment = ToolDeployment(*args, **kwargs)
+        tool_deployment.save()
+        return tool_deployment
+
+    def filter(self, **kwargs):
+        user = kwargs.get('user')
+        filter = Q(chart_name=None)  # Always False
+        for deployment in cluster.list_tool_deployments(user):
+            _, version = deployment.metadata.labels["chart"].rsplit("-", 1)
+            filter = filter | (
+                Q(chart_name=deployment.metadata.labels["app"]) &
+                Q(version=version)
+            )
+
+        tools = Tool.objects.filter(filter)
+        return [ToolDeployment(tool, user) for tool in tools]
+
+
+class ToolDeployment:
+    """
+    Represents a deployed Tool in the cluster
+    """
+
+    DoesNotExist = django.core.exceptions.ObjectDoesNotExist
+    Error = cluster.ToolDeploymentError
+    MultipleObjectsReturned = django.core.exceptions.MultipleObjectsReturned
+
+    objects = ToolDeploymentManager()
+
+    def __init__(self, tool, user):
+        self._subprocess = None
+        self.tool = tool
+        self.user = user
+
+    def __repr__(self):
+        return f'<ToolDeployment: {self.tool!r} {self.user!r}>'
+
+    def delete(self):
+        """
+        Remove the release from the cluster
+        """
+        cluster.delete_tool_deployment(self)
+
+    @property
+    def host(self):
+        return f'{self.user.slug}-{self.tool.chart_name}.{settings.TOOLS_DOMAIN}'
```

Comment:
It seems to work in my simple case but I notice that this is slightly different compared to what we had before: https://github.com/ministryofjustice/analytics-platform-control-panel/pull/732/files#diff-aad7b41260663fb400ff942a02fd3d14L200

```Python
    @property	
    def name(self):	
        return self.deployment.metadata.labels["app"]	

    @property	
    def username(self):	
        return self.user.username.lower()	

    @property	
    def url(self):	
        return f"https://{self.username}-{self.name}.{settings.TOOLS_DOMAIN}"
```

I wonder if there is some weird edge-case lurking in here? 

**Nevermind** it seems like the `app` label in the deployments is actually just `{{{ .Chart.Name }}`. I'm mainly wondering what would happen if we decide to allow 2 major/supported versions of a tool to run side-by-side. But it seems like a problem for another day which we may never have or what to support.

Reply:
Yeah, this is what we were talking about with having to put version numbers in the hostname or maybe using the randomly generated release names instead. I think we should standardise the hostnames - this particular pattern is not necessarily the right one, but we currently have two different ones (`username-rstudio` and `jupyter-lab-username`) which sucks.